Improve charts accessibility

Aljullu opened the pull request:
This PR implements most (if not all) changes suggested in #208.

* #339 made it possible to focus columns (dates) in charts, this PR makes it possible to focus every single point (in line charts) and every single bar (in bar charts).

* Each point/bar has an `aria-label` property used by screen readers with the date, the label and the value.

* The Y and X axis are hidden. Given that the `aria-label` of each point already includes that information, we don't need them to be visible to screen readers.

* Functions `handleMouseOverBarChart`, `handleMouseOutLineChart`, etc. have been modified a bit, so we pass them fewer parameters. And I also removed some `mouseover`, `mouseout` events that were never triggered.

**Screenshot**
![image](https://user-images.githubusercontent.com/3616980/45628338-272d3380-ba94-11e8-9b71-31640bd3c302.png)

**Steps to test**
* Activate the Screen Reader in your system. That depends on the operating system you're using, in Gnome Linux distros it is under _Settings_ > _Universal access_.
* Go to any page with a chart.
* Navigate the page with the `Tab` key.
* Verify that you can focus chart elements and, when focused, the tooltip is displayed.
* Verify that the screen reader reads out the chart information.
* Check both, the line chart and the bar chart.

**Future work**
* The buttons to change between line chart and bar chart are not accessible (#422).
![image](https://user-images.githubusercontent.com/3616980/45628233-efbe8700-ba93-11e8-9430-19b74b81ac8e.png)
* For now I'm using the tooltip date for the screen reader, but with #373 I understand it will disappear. So a new param will need to be created (or renaming the `tooltipFormat` one).

Disclaimer: I'm not very experienced in accessibility, so feel free to do a harsh review and point me out any issue you encounter! :slightly_smiling_face:

Aljullu committed:
Add aria roles to chart elements

Aljullu committed:
Make individual points (in line charts) and individual bar (in bar charts) focusable

Aljullu committed:
Remove methods which are never used

Aljullu committed:
Reduce the number of parameters required by functions that display/hide the tooltip

Aljullu committed:
Use tooltipFormat for accessibility dates

Aljullu committed:
Rename 'formatVoiceDate' function to 'getTooltipDate'

Aljullu committed:
Use string literals for aria-label

Aljullu committed:
Remove table role which was no longer needed

timmyc reviewed (commented):
Nice work here! I can navigate the chart tabs and voiceover reads out the active data point for me. The only odd thing I noticed is when on a point in the line chart, it would say "You are on an image" - wonder if we should have it say something else, like you are on a point in the chart perhaps?

I'll let @ryelle give this some 👀 too

ryelle commented on client/components/chart/utils.js:
```diff
@@ -493,10 +502,16 @@ export const drawLines = ( node, data, params ) => {
+			.attr( 'tabindex', '0' )
+			.attr(
+				'aria-label',
+				d => `${ getTooltipDate( params, d.date ) } ${ d.key } ${ formatCurrency( d.value ) }`
```
Having `key` in this label makes the label too long & a bit confusing on reports, for example I hear: `August 06 2018 Last Month (Aug 1 - 31, 2018) $8,080.81` – it's unclear what that dollar value represents, the entire month or just that day? Since each line is in a region with the key label, I think we can omit it here.

ryelle commented on client/components/chart/utils.js:
```diff
@@ -573,14 +586,20 @@ export const drawBars = ( node, data, params ) => {
+		.attr( 'tabindex', '0' )
+		.attr(
+			'aria-label',
+			d => `${ getTooltipDate( params, d.date ) } ${ d.key } ${ formatCurrency( d.value ) }`
```
I have the same problem as above, the key is extra info & can be omitted

ryelle replied:
I see/hear that this is a problem with the previous period, since `getTooltipDate( params, d.date )` will always get the "current" date, not the date of the previous period's value. Is there a way to change that? So that if I'm comparing August 1-31 vs July 1-31, it can tell me `August 6th, 2018 $500` and then `July 6th, 2018 $450`?

Currently it would say `August 6 2018 Last Month (Aug 1 - 31, 2018) $500`, then `August 6 2018 Previous Period (Jul 1 - 31, 2018) $450`.

ryelle reviewed (commented):
This is a good start on screen reader compatibility 👍  I have a few comments on what the labels themselves should be, but moving through with VoiceOver worked fairly well.

I did notice that the x-axis is not hidden, so I still need to navigate through that before getting into the chart content.

Aljullu committed:
Add aria-hidden to X-axis in charts

Aljullu committed:
Remove 'key' from points/bar aria-label in charts

timmyc commented:
@Aljullu do you need another review pass on this PR?

Aljullu commented:
> @Aljullu do you need another review pass on this PR?

@timmyc not yet! I'm still working on this comment from @ryelle (https://github.com/woocommerce/wc-admin/pull/421#discussion_r218251012).

The code needed to do that is very related to issue #373, so I'm working on both things together. I will ping you again when it's ready. :slightly_smiling_face:

Aljullu committed:
Set different ARIA properties depending on chart mode (time or item comparison)

Aljullu commented:
@timmyc @ryelle I rebased the branch with the last changes in `master` and implemented the changes discussed in this PR.

Can you take a look when you have a moment?

Aljullu committed:
Label should default to an empty string instead of a 0

Aljullu commented:
Given that #467 adds some changes that will conflict with this branch, I will leave this one on hold until I merge that one and will ask your review again later.

Aljullu committed:
Use date format from params instead of hardcoded

Aljullu commented:
It's ready to be reviewed again @timmyc @ryelle. :slightly_smiling_face:

ryelle reviewed (approved):
Overall this is a great improvement. I still think the exact labels need updating to be more clear, but that should be done in a follow-up PR after some discussion about what's useful (maybe we can set aside some time to talk about analytics accessibility at the GM team day @timmyc?)

timmyc commented:
> maybe we can set aside some time to talk about analytics accessibility at the GM team day?

💯
